# Import libraries


In [1]:
import numpy as np
import math
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Import file

In [2]:

xlsx_file_path = '../data/raw/project_file.xlsx'

sheet_name = 'Data' 

data = pd.read_excel(xlsx_file_path, sheet_name=sheet_name)


# Visualize data

In [3]:
data.shape

(28259, 122)

In [4]:
data.head()

Country Code Country_ID                     Product_ID Product Code  \
0             8    Albania  Roundwood (wood in the rough)            1   
1             8    Albania  Roundwood (wood in the rough)            1   
2             8    Albania  Roundwood (wood in the rough)            1   
3             8    Albania  Roundwood (wood in the rough)            1   
4             8    Albania  Roundwood (wood in the rough)            1   

       Flow      Unit   1964D 1964S   1965D 1965S  ...    2017D 2017S  \
0   EXPORTS   1000 m3     NaN   NaN     NaN   NaN  ...    26.67     C   
1   EXPORTS  1000 US$     NaN   NaN     NaN   NaN  ...  1676.93     C   
2   IMPORTS   1000 m3     NaN   NaN     NaN   NaN  ...    30.47     C   
3   IMPORTS  1000 US$     NaN   NaN     NaN   NaN  ...  1648.96     C   
4  REMOVALS   1000 m3  2025.0     C  2060.0     C  ...  1230.01     C   

     2018D 2018S    2019D 2019S    2020D 2020S    2021D 2021S  
0    15.35     C    14.02     C    11.80     C    14.26     C  
1  1907.49     C  1804.94     C  1692.82     C  2214.79     C  
2    23.80     C    18.46     C    39.04     C    23.37     C  
3  2325.06     C  1548.85     C  3807.80     C  2199.58     C  
4   905.34     C   917.84     C   874.45     C  1021.91     C  

[5 rows x 122 columns]

In [5]:
data.tail()

Country Code             Country_ID  \
28254           891  Serbia and Montenegro   
28255           891  Serbia and Montenegro   
28256           891  Serbia and Montenegro   
28257           891  Serbia and Montenegro   
28258           891  Serbia and Montenegro   

                                              Product_ID Product Code  \
28254                   Non-coniferous sawnwood of birch   6.NC.Birch   
28255  Non-coniferous non-tropical sawnwood other tha...   6.NC.Other   
28256  Non-coniferous non-tropical sawnwood other tha...   6.NC.Other   
28257  Non-coniferous non-tropical sawnwood other tha...   6.NC.Other   
28258  Non-coniferous non-tropical sawnwood other tha...   6.NC.Other   

          Flow      Unit  1964D 1964S  1965D 1965S  ...  2017D 2017S  2018D  \
28254  IMPORTS  1000 US$    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   
28255  EXPORTS   1000 m3    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   
28256  EXPORTS  1000 US$    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   
28257  IMPORTS   1000 m3    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   
28258  IMPORTS  1000 US$    NaN   NaN    NaN   NaN  ...    NaN   NaN    NaN   

      2018S  2019D 2019S  2020D 2020S  2021D 2021S  
28254   NaN    NaN   NaN    NaN   NaN    NaN   NaN  
28255   NaN    NaN   NaN    NaN   NaN    NaN   NaN  
28256   NaN    NaN   NaN    NaN   NaN    NaN   NaN  
28257   NaN    NaN   NaN    NaN   NaN    NaN   NaN  
28258   NaN    NaN   NaN    NaN   NaN    NaN   NaN  

[5 rows x 122 columns]

# DATA CLEANING AND WRANGLING


In [6]:
#get columns names
data.columns

Index(['Country Code', 'Country_ID', 'Product_ID', 'Product Code', 'Flow',
       'Unit', '1964D', '1964S', '1965D', '1965S',
       ...
       '2017D', '2017S', '2018D', '2018S', '2019D', '2019S', '2020D', '2020S',
       '2021D', '2021S'],
      dtype='object', length=122)

some definitions are given as

UNECE/FAO TIMBER database, 1964-2021, as of March 2022. 

Data point (yearD), status (yearS). 

Status

C=calculated (sum of subitems) 

E1=technical estimate (based on similar data) 

E2=analyst estimate (based on outside source) 

E3=estimate based on official aggregate - R=repeated - N=national estimate - blank=official

In [7]:
#define function  to get unique values in a column

def unique_values(column):
    return data[column].unique()


In [8]:
unique_values('Country_ID')

array(['Albania', 'Andorra', 'Azerbaijan', 'Austria', 'Armenia',
       'Belgium', 'Belgium & Luxembourg', 'Bosnia and Herzegovina',
       'Bulgaria', 'Belarus', 'Canada', 'Croatia', 'Cyprus',
       'Former CSFR', 'Czechia', 'Denmark', 'Estonia', 'Finland',
       'France', 'Georgia', 'Germany',
       'Former German Democratic Republic',
       'Germany, Former Federal Republic of', 'Greece', 'Hungary',
       'Iceland', 'Ireland', 'Israel', 'Italy', 'Kazakhstan',
       'Kyrgyzstan', 'Latvia', 'Liechtenstein', 'Lithuania', 'Luxembourg',
       'Malta', 'Republic of Moldova', 'Montenegro', 'Netherlands',
       'Norway', 'Poland', 'Portugal', 'Romania', 'Russian Federation',
       'Serbia', 'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Switzerland',
       'Tajikistan', 'Turkey', 'Turkmenistan', 'Ukraine',
       'North Macedonia', 'Former Soviet Union', 'United Kingdom',
       'United States', 'Uzbekistan', 'Former SFRY',
       'Serbia and Montenegro'], dtype=object)

# DEALING WITH COUNTRIES


for convineience of analysis, our focus will be on classsic european nation 
Russia is problematic for our analysis as most of her older data are contained in the soviet union, and unless we aggregate the data of other constuitent states that composes the defunct USSR (see link, https://en.wikipedia.org/wiki/Post-Soviet_states), we would introduce errors if we are to include the data from russia .
so for convinient sake we will exclude Russia , future project might require using future estimates from russia and other post-soviet states to determine 
The post soviet states include
 1.  Armenia •
 2.  Azerbaijan • 
 3.  Belarus •
 4.  Estonia •
 5.  Georgia • 
 6.  Kazakhstan • 
 7.  Kyrgyzstan • 
 8.  Latvia • 
 9.  Lithuania • 
 10. Moldova • 
 11. Russia • 
 12. Tajikistan •
 13. Turkmenistan • 
 14. Ukraine 
 15. Uzbekistan

for other defunct states like serbia and montenegro, Czech and Slovak Federative Republic (CSFR),The Socialist Federal Republic of Yugoslavia (SFRY) we will also drop the countries for this analysis, a future possibilty  might require using them
so by that , we will drop mordern states like 


and also we will create a seperate data set that contain mordern data where only mordern states are included and use it answer question 2 . 

germany was split before 1993 to west(Germany, Former Federal Republic of) and east (Former German Democratic Republic')

we will concatenate both rows in a new data set and include as simply germany. Historically , the two partitioned german state has not lost any territory before and after reunification. So we won´t introduce any error to our data

In [9]:
unique_values('Product_ID')

array(['Roundwood (wood in the rough)', 'Coniferous roundwood',
       'Non-Coniferous roundwood',
       'Wood fuel, including wood for charcoal', 'Coniferous wood fuel',
       'Non-Coniferous wood fuel', 'Industrial roundwood',
       'Coniferous industrial roundwood',
       'Non-Coniferous industrial roundwood',
       'Tropical industrial roundwood',
       'Non-Coniferous non-tropical industrial roundwood',
       'Sawlogs and veneer logs', 'Coniferous sawlogs and veneer logs',
       'Non-Coniferous sawlogs and veneer logs',
       'Pulpwood (round & split)', 'Coniferous pulpwood',
       'Non-Coniferous Pulpwood', 'Other industrial roundwood',
       'Coniferous other industrial roundwood',
       'Non-Coniferous other industrial roundwood', 'Wood charcoal',
       'Wood residues, chips and particles', 'Chips and particles',
       'Wood residues', 'Recovered post-consumer wood',
       'Wood pellets and other agglomerates of wood', 'Wood pellets',
       'Other agglomerates o

In [10]:
#number of unique values in a column
def unique_counts(column):
    return data[column].nunique()

unique_counts('Product_ID')


130

Most efforts wil be centered around product_id column, we will reduce the variability by removing the sub-categories that tries to distinguish between coniferous, non-coniferous and others. For a study on forest ecosystem dynamism, that might be a useful categorization, but not for this study. 

Also all categories of paper productrs and packaging product will be categorized as paper and packaging
At the end we will have only paper and packaging.
All kinds of pulp will be joined, so it won´t matter if it´s kraft, sulfite , bleached or unbleached, virgin or recovered fibre 

All engineering boards (wood based panels) like veneer, plywood, osb, hardboard, MDF/HDF, particleboard,otherboards etc will be grouped under engineering board.

wood pellets and other agglomerate will be grouped together
wood residues, chips and pellets will be gropued together too
all pulpwood will be grouped 
.
At the end we will significantly reduce product category from 130 to less than 20.

In [11]:
#lets see what a row that has Roundwood (wood in the rough) looks like
data[data['Product_ID'] == 'Roundwood (wood in the rough)'].head()

Country Code Country_ID                     Product_ID Product Code  \
0             8    Albania  Roundwood (wood in the rough)            1   
1             8    Albania  Roundwood (wood in the rough)            1   
2             8    Albania  Roundwood (wood in the rough)            1   
3             8    Albania  Roundwood (wood in the rough)            1   
4             8    Albania  Roundwood (wood in the rough)            1   

       Flow      Unit   1964D 1964S   1965D 1965S  ...    2017D 2017S  \
0   EXPORTS   1000 m3     NaN   NaN     NaN   NaN  ...    26.67     C   
1   EXPORTS  1000 US$     NaN   NaN     NaN   NaN  ...  1676.93     C   
2   IMPORTS   1000 m3     NaN   NaN     NaN   NaN  ...    30.47     C   
3   IMPORTS  1000 US$     NaN   NaN     NaN   NaN  ...  1648.96     C   
4  REMOVALS   1000 m3  2025.0     C  2060.0     C  ...  1230.01     C   

     2018D 2018S    2019D 2019S    2020D 2020S    2021D 2021S  
0    15.35     C    14.02     C    11.80     C    14.26     C  
1  1907.49     C  1804.94     C  1692.82     C  2214.79     C  
2    23.80     C    18.46     C    39.04     C    23.37     C  
3  2325.06     C  1548.85     C  3807.80     C  2199.58     C  
4   905.34     C   917.84     C   874.45     C  1021.91     C  

[5 rows x 122 columns]

# CLEANING FLOW

In [12]:
unique_values('Flow')

array(['EXPORTS', 'IMPORTS', 'REMOVALS', 'PRODUCTION'], dtype=object)

To answer our first question; REMOVALS will be regarded as harvest,
trade deficit will be looked into comparing import and exports 
#Production cannot be a good determinant and that might also include locally and imported raw materials #

In [13]:
unique_values('2015S')

array(['C', 'R', nan, 'E1', 'E2', 'N', 'E3'], dtype=object)

# CLEANING THE DATA

OBJECTIVE 1
-clean up product 
how- define functions to  identify keywords in product_id and group them on a new column named named product
at the end we will have products reduced to  about 30 products categories
A lot of products will be classified as others , in a future analysis, these products will be analyzed on individual merit


In [14]:
#make all values in column product_id lowercase
data['Product_ID'] = data['Product_ID'].str.lower()

In [15]:
#,define function to create a new column which will be made by checking the character in a column and assigninga string value to that character which will form the new column
def map_new_column(value):
    '''
    This function creates a new column based on the character in a column
    param char: character in a column
    output: new column with string values
    '''
    if 'roundwood' in value:
        return 'roundwood'
    elif 'wood fuel' in value:
        return 'wood fuel'
    elif 'sawlogs and veneer logs' in value:
        return 'sawlogs and veneer logs'
    elif 'pulpwood' in value :
        return 'pulpwood'
    elif 'wood pulp' in value or 'Wood pulp' in value:
        return 'wood pulp'
    elif 'industrial roundwood' in value:
        return 'industrial roundwood'
    elif 'wood residue' in value:
        return 'wood residue, chips and particles'
    elif 'chips and particles' in value:
        return 'wood residue, chips and particles'
    elif 'wood pellets' in value or 'other wood agglomerates' in value or 'other agglomerates of wood' in value:
        return 'wood pellets and other wood agglomerates'
    elif  'sawnwood' in value:
        return 'sawnwood'
    elif 'veneer sheets' in value:
        return 'veneer sheets'
    elif 'wood-based panels' in value: 
        return 'wood-based panels'
    elif 'plywood' in value or 'particle board' in value or 'osb' in value or 'fibreboard' in value or 'hardboard' in value or 'other board' in value:
        return 'engineered wood products'
    elif 'woodpulp' in value:
        return 'woodpulp'
    elif 'pulp from fibres other than wood' in value:
        return 'non-wood pulp'
    elif 'paper' in value or 'papers' in value or 'packaging' in value or 'cartonboard' in value or 'paperboard' in value or 'printed' in value or 'newsprint' in value:
        return 'paper,print and packaging'
    elif 'domestic' in value:
        return 'domestic/decorative products'
    elif 'sleepers' in value:
        return 'sleepers'
    else:
        return value





In [16]:
#apply function to create new column
data['product'] = data['Product_ID'].apply(map_new_column)

In [17]:
data['product'].unique()

array(['roundwood', 'wood fuel', 'sawlogs and veneer logs', 'pulpwood',
       'wood charcoal', 'wood residue, chips and particles',
       'recovered post-consumer wood',
       'wood pellets and other wood agglomerates', 'sawnwood',
       'wood-based panels', 'veneer sheets', 'engineered wood products',
       'woodpulp', 'other pulp', 'non-wood pulp', 'recovered fibre pulp',
       'paper,print and packaging', 'case materials',
       'secondary processed wooden products',
       'wooden wrapping and packing equipment',
       'domestic/decorative products', 'other manufactured wood products',
       "builder's joinery and carpentry of wood", 'wooden furniture',
       'prefabricated buildings (discontinued 2017)',
       'prefabricated buildings of wood',
       'articles, moulded or pressed from pulp', 'sleepers'], dtype=object)

In [18]:
data['product'].value_counts()

product
paper,print and packaging                      5561
roundwood                                      4350
sawnwood                                       3742
woodpulp                                       3430
engineered wood products                       3316
veneer sheets                                  1113
wood residue, chips and particles               863
wood fuel                                       819
wood pellets and other wood agglomerates        778
pulpwood                                        716
wood-based panels                               595
sawlogs and veneer logs                         374
other pulp                                      284
non-wood pulp                                   284
wood charcoal                                   277
recovered fibre pulp                            273
case materials                                  273
recovered post-consumer wood                    178
sleepers                                        123
seco

In [19]:
data['product'].nunique()

28

# CLEANING YEAR COLUMN
Due to the nature of our analysis, the yearStatus column will not be put into consideration. 
Status
C=calculated (sum of subitems) 

E1=technical estimate (based on similar data) 

E2=analyst estimate (based on outside source) 

E3=estimate based on official aggregate 

R=repeated 

N=national estimate 

blank=official

so we are more concerned with the value than the way it was calculated ,as we assume that the calculation method used is reliable or at least approximate to actual value
As a result, we will drop the date column with S a a suffix , meaning Status

In [20]:
import re

columns_to_drop = [column for column in list(data.columns) if re.match("\d{4}S", column)]
data.drop(columns=columns_to_drop, inplace=True)

In [21]:
data.head()

Country Code Country_ID                     Product_ID Product Code  \
0             8    Albania  roundwood (wood in the rough)            1   
1             8    Albania  roundwood (wood in the rough)            1   
2             8    Albania  roundwood (wood in the rough)            1   
3             8    Albania  roundwood (wood in the rough)            1   
4             8    Albania  roundwood (wood in the rough)            1   

       Flow      Unit   1964D   1965D   1966D   1967D  ...    2013D    2014D  \
0   EXPORTS   1000 m3     NaN     NaN     NaN     NaN  ...    78.90    78.90   
1   EXPORTS  1000 US$     NaN     NaN     NaN     NaN  ...  3830.00  3830.00   
2   IMPORTS   1000 m3     NaN     NaN     NaN     NaN  ...   105.00   120.00   
3   IMPORTS  1000 US$     NaN     NaN     NaN     NaN  ...  3741.00  4686.00   
4  REMOVALS   1000 m3  2025.0  2060.0  2135.0  2112.0  ...  1180.01  1180.01   

     2015D    2016D    2017D    2018D    2019D    2020D    2021D    product  
0    78.90    78.90    26.67    15.35    14.02    11.80    14.26  roundwood  
1  3830.00  3830.00  1676.93  1907.49  1804.94  1692.82  2214.79  roundwood  
2   120.00   120.00    30.47    23.80    18.46    39.04    23.37  roundwood  
3  4686.00  4686.00  1648.96  2325.06  1548.85  3807.80  2199.58  roundwood  
4  1180.01  1180.01  1230.01   905.34   917.84   874.45  1021.91  roundwood  

[5 rows x 65 columns]

now first we will create a new data frame titled data_with_grp 

In [22]:
#create a new dataframe that replaces product_id with the new column and drops all column with S as last character

data_new = data.copy()
data_new.drop(columns=['Product_ID', 'Product Code'], inplace=True)


In [23]:
data_new.head()

Country Code Country_ID      Flow      Unit   1964D   1965D   1966D  \
0             8    Albania   EXPORTS   1000 m3     NaN     NaN     NaN   
1             8    Albania   EXPORTS  1000 US$     NaN     NaN     NaN   
2             8    Albania   IMPORTS   1000 m3     NaN     NaN     NaN   
3             8    Albania   IMPORTS  1000 US$     NaN     NaN     NaN   
4             8    Albania  REMOVALS   1000 m3  2025.0  2060.0  2135.0   

    1967D   1968D   1969D  ...    2013D    2014D    2015D    2016D    2017D  \
0     NaN     NaN     NaN  ...    78.90    78.90    78.90    78.90    26.67   
1     NaN     NaN     NaN  ...  3830.00  3830.00  3830.00  3830.00  1676.93   
2     NaN     NaN     NaN  ...   105.00   120.00   120.00   120.00    30.47   
3     NaN     NaN     NaN  ...  3741.00  4686.00  4686.00  4686.00  1648.96   
4  2112.0  2200.0  2320.0  ...  1180.01  1180.01  1180.01  1180.01  1230.01   

     2018D    2019D    2020D    2021D    product  
0    15.35    14.02    11.80    14.26  roundwood  
1  1907.49  1804.94  1692.82  2214.79  roundwood  
2    23.80    18.46    39.04    23.37  roundwood  
3  2325.06  1548.85  3807.80  2199.58  roundwood  
4   905.34   917.84   874.45  1021.91  roundwood  

[5 rows x 63 columns]

In [24]:
data_new.columns

Index(['Country Code', 'Country_ID', 'Flow', 'Unit', '1964D', '1965D', '1966D',
       '1967D', '1968D', '1969D', '1970D', '1971D', '1972D', '1973D', '1974D',
       '1975D', '1976D', '1977D', '1978D', '1979D', '1980D', '1981D', '1982D',
       '1983D', '1984D', '1985D', '1986D', '1987D', '1988D', '1989D', '1990D',
       '1991D', '1992D', '1993D', '1994D', '1995D', '1996D', '1997D', '1998D',
       '1999D', '2000D', '2001D', '2002D', '2003D', '2004D', '2005D', '2006D',
       '2007D', '2008D', '2009D', '2010D', '2011D', '2012D', '2013D', '2014D',
       '2015D', '2016D', '2017D', '2018D', '2019D', '2020D', '2021D',
       'product'],
      dtype='object')


# Rearrange table to change the position 

we will now change the order of the to include new column 'product in row index 3

In [25]:
#rearraange columns in new dataframe 

data_new = data_new[[ 'Country Code','Country_ID', 'Flow', 'product', 'Unit', '1964D', '1965D', '1966D',
       '1967D', '1968D', '1969D', '1970D', '1971D', '1972D', '1973D', '1974D',
       '1975D', '1976D', '1977D', '1978D', '1979D', '1980D', '1981D', '1982D',
       '1983D', '1984D', '1985D', '1986D', '1987D', '1988D', '1989D', '1990D',
       '1991D', '1992D', '1993D', '1994D', '1995D', '1996D', '1997D', '1998D',
       '1999D', '2000D', '2001D', '2002D', '2003D', '2004D', '2005D', '2006D',
       '2007D', '2008D', '2009D', '2010D', '2011D', '2012D', '2013D', '2014D',
       '2015D', '2016D', '2017D', '2018D', '2019D', '2020D', '2021D']]

In [26]:
data_new.head()

Country Code Country_ID      Flow    product      Unit   1964D   1965D  \
0             8    Albania   EXPORTS  roundwood   1000 m3     NaN     NaN   
1             8    Albania   EXPORTS  roundwood  1000 US$     NaN     NaN   
2             8    Albania   IMPORTS  roundwood   1000 m3     NaN     NaN   
3             8    Albania   IMPORTS  roundwood  1000 US$     NaN     NaN   
4             8    Albania  REMOVALS  roundwood   1000 m3  2025.0  2060.0   

    1966D   1967D   1968D  ...    2012D    2013D    2014D    2015D    2016D  \
0     NaN     NaN     NaN  ...    78.90    78.90    78.90    78.90    78.90   
1     NaN     NaN     NaN  ...  3830.00  3830.00  3830.00  3830.00  3830.00   
2     NaN     NaN     NaN  ...     6.38   105.00   120.00   120.00   120.00   
3     NaN     NaN     NaN  ...   743.00  3741.00  4686.00  4686.00  4686.00   
4  2135.0  2112.0  2200.0  ...  1180.00  1180.01  1180.01  1180.01  1180.01   

     2017D    2018D    2019D    2020D    2021D  
0    26.67    15.35    14.02    11.80    14.26  
1  1676.93  1907.49  1804.94  1692.82  2214.79  
2    30.47    23.80    18.46    39.04    23.37  
3  1648.96  2325.06  1548.85  3807.80  2199.58  
4  1230.01   905.34   917.84   874.45  1021.91  

[5 rows x 63 columns]

In [27]:
#drop D behind year column names but ignore country_id
modify_columns = lambda x: 'Country_ID' if x == 'Country_ID' else x.replace('D', '')
data_new.columns = data_new.columns.map(modify_columns)

In [28]:
data_new.head()

Country Code Country_ID      Flow    product      Unit    1964    1965  \
0             8    Albania   EXPORTS  roundwood   1000 m3     NaN     NaN   
1             8    Albania   EXPORTS  roundwood  1000 US$     NaN     NaN   
2             8    Albania   IMPORTS  roundwood   1000 m3     NaN     NaN   
3             8    Albania   IMPORTS  roundwood  1000 US$     NaN     NaN   
4             8    Albania  REMOVALS  roundwood   1000 m3  2025.0  2060.0   

     1966    1967    1968  ...     2012     2013     2014     2015     2016  \
0     NaN     NaN     NaN  ...    78.90    78.90    78.90    78.90    78.90   
1     NaN     NaN     NaN  ...  3830.00  3830.00  3830.00  3830.00  3830.00   
2     NaN     NaN     NaN  ...     6.38   105.00   120.00   120.00   120.00   
3     NaN     NaN     NaN  ...   743.00  3741.00  4686.00  4686.00  4686.00   
4  2135.0  2112.0  2200.0  ...  1180.00  1180.01  1180.01  1180.01  1180.01   

      2017     2018     2019     2020     2021  
0    26.67    15.35    14.02    11.80    14.26  
1  1676.93  1907.49  1804.94  1692.82  2214.79  
2    30.47    23.80    18.46    39.04    23.37  
3  1648.96  2325.06  1548.85  3807.80  2199.58  
4  1230.01   905.34   917.84   874.45  1021.91  

[5 rows x 63 columns]

time to rename columns

In [29]:
#make column names lowercase and replace space with underscore
data_new.columns = data_new.columns.str.lower().str.replace(' ', '_')
data_new.columns

Index(['country_code', 'country_id', 'flow', 'product', 'unit', '1964', '1965',
       '1966', '1967', '1968', '1969', '1970', '1971', '1972', '1973', '1974',
       '1975', '1976', '1977', '1978', '1979', '1980', '1981', '1982', '1983',
       '1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991', '1992',
       '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001',
       '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010',
       '2011', '2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019',
       '2020', '2021'],
      dtype='object')

In [30]:
#assgn new product code to product column and create a new column for product code
data_new['product_code'] = data_new['product'].astype('category').cat.codes

In [31]:
#select product and product code columns
data_new[['product', 'product_code']].head(60)


product  product_code
0                 roundwood            14
1                 roundwood            14
2                 roundwood            14
3                 roundwood            14
4                 roundwood            14
5                 roundwood            14
6                 roundwood            14
7                 roundwood            14
8                 roundwood            14
9                 roundwood            14
10                roundwood            14
11                roundwood            14
12                roundwood            14
13                roundwood            14
14                roundwood            14
15                wood fuel            21
16                wood fuel            21
17                wood fuel            21
18                wood fuel            21
19                wood fuel            21
20                wood fuel            21
21                wood fuel            21
22                wood fuel            21
23                wood fuel            21
24                wood fuel            21
25                wood fuel            21
26                wood fuel            21
27                wood fuel            21
28                wood fuel            21
29                wood fuel            21
30                roundwood            14
31                roundwood            14
32                roundwood            14
33                roundwood            14
34                roundwood            14
35                roundwood            14
36                roundwood            14
37                roundwood            14
38                roundwood            14
39                roundwood            14
40                roundwood            14
41                roundwood            14
42                roundwood            14
43                roundwood            14
44                roundwood            14
45                roundwood            14
46                roundwood            14
47                roundwood            14
48                roundwood            14
49                roundwood            14
50                roundwood            14
51                roundwood            14
52                roundwood            14
53                roundwood            14
54                roundwood            14
55  sawlogs and veneer logs            15
56  sawlogs and veneer logs            15
57  sawlogs and veneer logs            15
58                 pulpwood            11
59                 pulpwood            11

In [32]:
#change the order of columns , so that product_code is next to product

data_new = data_new[[ 'country_code','country_id', 'product', 'product_code','flow', 'unit', '1964', '1965', '1966', '1967', '1968', '1969', '1970', '1971', '1972', '1973', '1974','1975', '1976', '1977', '1978', '1979', '1980', '1981', '1982', '1983','1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991', '1992','1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001','2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010','2011', '2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019','2020', '2021']]

In [33]:
data_new.head()

country_code country_id    product  product_code      flow      unit  \
0             8    Albania  roundwood            14   EXPORTS   1000 m3   
1             8    Albania  roundwood            14   EXPORTS  1000 US$   
2             8    Albania  roundwood            14   IMPORTS   1000 m3   
3             8    Albania  roundwood            14   IMPORTS  1000 US$   
4             8    Albania  roundwood            14  REMOVALS   1000 m3   

     1964    1965    1966    1967  ...     2012     2013     2014     2015  \
0     NaN     NaN     NaN     NaN  ...    78.90    78.90    78.90    78.90   
1     NaN     NaN     NaN     NaN  ...  3830.00  3830.00  3830.00  3830.00   
2     NaN     NaN     NaN     NaN  ...     6.38   105.00   120.00   120.00   
3     NaN     NaN     NaN     NaN  ...   743.00  3741.00  4686.00  4686.00   
4  2025.0  2060.0  2135.0  2112.0  ...  1180.00  1180.01  1180.01  1180.01   

      2016     2017     2018     2019     2020     2021  
0    78.90    26.67    15.35    14.02    11.80    14.26  
1  3830.00  1676.93  1907.49  1804.94  1692.82  2214.79  
2   120.00    30.47    23.80    18.46    39.04    23.37  
3  4686.00  1648.96  2325.06  1548.85  3807.80  2199.58  
4  1180.01  1230.01   905.34   917.84   874.45  1021.91  

[5 rows x 64 columns]

Now we will save this file that has reduced product category for future use 

In [34]:
#save new dataframe to csv
data_new.to_csv('../data/raw/product_cat.csv', index=False)

# DEALING WITH COUNTRIES

In [35]:
#use producr_cat.csv 
data2 = pd.read_csv('../data/raw/product_cat.csv')

In [36]:
data2.head()

country_code country_id    product  product_code      flow      unit  \
0             8    Albania  roundwood            14   EXPORTS   1000 m3   
1             8    Albania  roundwood            14   EXPORTS  1000 US$   
2             8    Albania  roundwood            14   IMPORTS   1000 m3   
3             8    Albania  roundwood            14   IMPORTS  1000 US$   
4             8    Albania  roundwood            14  REMOVALS   1000 m3   

     1964    1965    1966    1967  ...     2012     2013     2014     2015  \
0     NaN     NaN     NaN     NaN  ...    78.90    78.90    78.90    78.90   
1     NaN     NaN     NaN     NaN  ...  3830.00  3830.00  3830.00  3830.00   
2     NaN     NaN     NaN     NaN  ...     6.38   105.00   120.00   120.00   
3     NaN     NaN     NaN     NaN  ...   743.00  3741.00  4686.00  4686.00   
4  2025.0  2060.0  2135.0  2112.0  ...  1180.00  1180.01  1180.01  1180.01   

      2016     2017     2018     2019     2020     2021  
0    78.90    26.67    15.35    14.02    11.80    14.26  
1  3830.00  1676.93  1907.49  1804.94  1692.82  2214.79  
2   120.00    30.47    23.80    18.46    39.04    23.37  
3  4686.00  1648.96  2325.06  1548.85  3807.80  2199.58  
4  1180.01  1230.01   905.34   917.84   874.45  1021.91  

[5 rows x 64 columns]

In [37]:
data2['country_id'].unique()

array(['Albania', 'Andorra', 'Azerbaijan', 'Austria', 'Armenia',
       'Belgium', 'Belgium & Luxembourg', 'Bosnia and Herzegovina',
       'Bulgaria', 'Belarus', 'Canada', 'Croatia', 'Cyprus',
       'Former CSFR', 'Czechia', 'Denmark', 'Estonia', 'Finland',
       'France', 'Georgia', 'Germany',
       'Former German Democratic Republic',
       'Germany, Former Federal Republic of', 'Greece', 'Hungary',
       'Iceland', 'Ireland', 'Israel', 'Italy', 'Kazakhstan',
       'Kyrgyzstan', 'Latvia', 'Liechtenstein', 'Lithuania', 'Luxembourg',
       'Malta', 'Republic of Moldova', 'Montenegro', 'Netherlands',
       'Norway', 'Poland', 'Portugal', 'Romania', 'Russian Federation',
       'Serbia', 'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Switzerland',
       'Tajikistan', 'Turkey', 'Turkmenistan', 'Ukraine',
       'North Macedonia', 'Former Soviet Union', 'United Kingdom',
       'United States', 'Uzbekistan', 'Former SFRY',
       'Serbia and Montenegro'], dtype=object)

like we mentioned earlier, we are concerned with the  in historical data, so we are going to drop defunct states like USSR, yugoslavia and serbia-montenegro, czechslovakia and states that emerged from the aforementioned states  and convert to a csv file. 
But given the uniqueness of clarity in territorial definition pre and post reunification, we will contanate west(Germany, Former Federal Republic of) and east (Former German Democratic Republic') data and use them as historical data for germany.  

For other comparisons like comparison of resource efficiency and volume vs value comparison, we will retain only records from the last 15 years (2007-2021). This means we will drop the defunct states by default and any state that ceased to exist after 2006 like serbia-montenegro
 

former sovie union countries are 
 Armenia ,Azerbaijan ,Belarus ,Estonia ,Georgia ,Kazakhstan  Kyrgyzstan, Latvia, Lithuania, Moldova, Russia ,Tajikistan,Turkmenistan, Ukraine ,Uzbekistan

former yugoslavia countries are Croatia, Montenegro, Serbia, Slovenia, Bosnia and Herzegovina, and Macedonia.

former czechoslovakia countries are Czech Republic and Slovakia

Serbi and Montenegro are now separate countries and so will be dropped

Belgium and Luxemborg too





first we will drop columns before 2006 for some of our questions

In [38]:
#locate col index 1964
data2.columns.get_loc('1964')

6

In [39]:
#locate col index 2006
data2.columns.get_loc('2007')

49

In [40]:
data3 = data2.copy()

In [41]:
#drop columns from 1964 to 2006
data3.drop(data3.iloc[:, 6:49], inplace=True, axis=1)
data3.head()

country_code country_id    product  product_code      flow      unit  \
0             8    Albania  roundwood            14   EXPORTS   1000 m3   
1             8    Albania  roundwood            14   EXPORTS  1000 US$   
2             8    Albania  roundwood            14   IMPORTS   1000 m3   
3             8    Albania  roundwood            14   IMPORTS  1000 US$   
4             8    Albania  roundwood            14  REMOVALS   1000 m3   

      2007     2008     2009     2010  ...     2012     2013     2014  \
0    56.76    56.76    56.76    56.76  ...    78.90    78.90    78.90   
1  1193.11  1193.11  1193.11  1193.11  ...  3830.00  3830.00  3830.00   
2     0.59     0.59     0.59     1.87  ...     6.38   105.00   120.00   
3    68.69    68.69    68.69   559.29  ...   743.00  3741.00  4686.00   
4   430.00   430.00   430.00   430.00  ...  1180.00  1180.01  1180.01   

      2015     2016     2017     2018     2019     2020     2021  
0    78.90    78.90    26.67    15.35    14.02    11.80    14.26  
1  3830.00  3830.00  1676.93  1907.49  1804.94  1692.82  2214.79  
2   120.00   120.00    30.47    23.80    18.46    39.04    23.37  
3  4686.00  4686.00  1648.96  2325.06  1548.85  3807.80  2199.58  
4  1180.01  1180.01  1230.01   905.34   917.84   874.45  1021.91  

[5 rows x 21 columns]

In [42]:
data3.shape

(28259, 21)

In [43]:
#save new dataframe to csv
data3.to_csv('../data/raw/data_06-21.csv', index=False)

The above data will be used to answer most questions , but we will obtain anew dataframe and csv file that excludes defunct states and their propsective breakout state while we concatanate  west(Germany, Former Federal Republic of) and east (Former German Democratic Republic') as germany
 

In [44]:
data4= data2.copy()

drop former soviet states :Armenia ,Azerbaijan ,Belarus ,Estonia ,Georgia ,Kazakhstan  Kyrgyzstan, Latvia, Lithuania, Moldova, Russia ,Tajikistan,Turkmenistan, Ukraine ,Uzbekistan

#make a list of former soviet states


#drop rows where country_id is in soviet_states
data2 = data2[~data2['country_id'].isin(soviet_states)]
data2['country_id'].unique()

In [45]:
#define a function that takes a column and list as an argument goes through the column and drop any row that has a value in the list

def drop_rows(df,column, lst):
    '''
    This function takes a column: country_id and A list (country)as an argument goes through the column and drop any row that has a value in the list
    param column: column to be checked
    param lst: list of values to be dropped
    output: dataframe with dropped rows
    '''
    return data4[~data4[column].isin(lst)]


In [46]:
#drop selected defunct states
defunct_states=['Former CSFR','Former Soviet Union','Former SFRY','Former CSFR']
data4 = drop_rows(data4,'country_id', defunct_states)

In [47]:
soviet_states = ['Armenia', 'Azerbaijan', 'Belarus', 'Estonia', 'Georgia', 'Kazakhstan', 'Kyrgyzstan', 'Latvia', 'Lithuania', 'Republic of Moldova', 'Russian Federation', 'Tajikistan', 'Turkmenistan', 'Ukraine', 'Uzbekistan']
lst= soviet_states
#call function to drop rows
data4=drop_rows(data4,'country_id', soviet_states)

In [48]:
data4['country_id'].unique()    

array(['Albania', 'Andorra', 'Austria', 'Belgium', 'Belgium & Luxembourg',
       'Bosnia and Herzegovina', 'Bulgaria', 'Canada', 'Croatia',
       'Cyprus', 'Czechia', 'Denmark', 'Finland', 'France', 'Germany',
       'Former German Democratic Republic',
       'Germany, Former Federal Republic of', 'Greece', 'Hungary',
       'Iceland', 'Ireland', 'Israel', 'Italy', 'Liechtenstein',
       'Luxembourg', 'Malta', 'Montenegro', 'Netherlands', 'Norway',
       'Poland', 'Portugal', 'Romania', 'Serbia', 'Slovakia', 'Slovenia',
       'Spain', 'Sweden', 'Switzerland', 'Turkey', 'North Macedonia',
       'United Kingdom', 'United States', 'Serbia and Montenegro'],
      dtype=object)

In [49]:
former_yugoslavia = ['Bosnia and Herzegovina', 'Croatia', 'Kosovo', 'Montenegro', 'North Macedonia', 'Serbia', 'Slovenia']

#call function to drop rows
data4=drop_rows(data4,'country_id', former_yugoslavia)

In [50]:
data4['country_id'].unique()

array(['Albania', 'Andorra', 'Austria', 'Belgium', 'Belgium & Luxembourg',
       'Bulgaria', 'Canada', 'Cyprus', 'Czechia', 'Denmark', 'Finland',
       'France', 'Germany', 'Former German Democratic Republic',
       'Germany, Former Federal Republic of', 'Greece', 'Hungary',
       'Iceland', 'Ireland', 'Israel', 'Italy', 'Liechtenstein',
       'Luxembourg', 'Malta', 'Netherlands', 'Norway', 'Poland',
       'Portugal', 'Romania', 'Slovakia', 'Spain', 'Sweden',
       'Switzerland', 'Turkey', 'United Kingdom', 'United States',
       'Serbia and Montenegro'], dtype=object)

In [51]:
#drop czechoslovakia
czechoslovakia = ['Czechia', 'Slovakia']
#call function to drop rows
data4=drop_rows(data4,'country_id', czechoslovakia)

In [52]:
#drop serbia and montenegro
serbia_montenegro = ['Serbia', 'Montenegro', 'Serbia and Montenegro']
#call function to drop rows
data4=drop_rows(data4,'country_id', serbia_montenegro)

In [53]:
#drop belgium-luxembourg
belgium_luxembourg = ['Belgium & Luxembourg', 'Belgium', 'Luxembourg']
#call function to drop rows
data4=drop_rows(data4,'country_id', belgium_luxembourg)

In [54]:
data4['country_id'].unique()

array(['Albania', 'Andorra', 'Austria', 'Bulgaria', 'Canada', 'Cyprus',
       'Denmark', 'Finland', 'France', 'Germany',
       'Former German Democratic Republic',
       'Germany, Former Federal Republic of', 'Greece', 'Hungary',
       'Iceland', 'Ireland', 'Israel', 'Italy', 'Liechtenstein', 'Malta',
       'Netherlands', 'Norway', 'Poland', 'Portugal', 'Romania', 'Spain',
       'Sweden', 'Switzerland', 'Turkey', 'United Kingdom',
       'United States'], dtype=object)

some recent countries like albania will be dropped based on their lack of historical data(Nan), but we will try to concatanate former german dem rep and former federal republic. and assign country_id as germany  

In [55]:
#now we export the data to csv
data4.to_csv('../data/raw/country_to_clean.csv', index=False)